src/vma/dev/net_device_table_mgr.cpp

```diff
@@ -323,10 +344,122 @@ void net_device_table_mgr::verify_ipoib_mode(struct ifaddrs* ifa)
 		vlog_printf(VLOG_WARNING,"Please refer to Release Notes for more information\n");
 		vlog_printf(VLOG_WARNING,"************************************************************************\n");
+		return false;
 	}
 	else {
 		ndtm_logdbg("verified interface '%s' is running with umcast disabled", ifa->ifa_name);
 	}
+	return true;
+}
+
+bool net_device_table_mgr::check_mlnx4_ib_device(const char* ifname)
+{
+	if (!check_device_exist(ifname, MLX4_DRIVER_PATH)) {
+		ndtm_logdbg("Flow Steering of IPoIB traffic is not supported on your HCA\n");
+		ndtm_logdbg("Please refer to VMA Release Notes for details limitations.\n");
+		ndtm_logdbg("All traffic over interface %s will not be offloaded!\n", ifname);
+		return false;
+	}
+	return true;
+}
+
+bool net_device_table_mgr::verify_qp_creation(const char* ifname, enum ibv_qp_type qp_type)
+{
+	int num_devices = 0;
+	bool success = false;
+
+	struct ibv_qp_init_attr qp_init_attr;
+	memset(&qp_init_attr, 0, sizeof(qp_init_attr));
+
+	vma_ibv_cq_init_attr attr;
+	memset(&attr, 0, sizeof(attr));
+
+	qp_init_attr.cap.max_send_wr = safe_mce_sys().tx_num_wr;
+	qp_init_attr.cap.max_recv_wr = safe_mce_sys().rx_num_wr;
+	qp_init_attr.cap.max_inline_data = safe_mce_sys().tx_max_inline;
+	qp_init_attr.cap.max_send_sge = MCE_DEFAULT_TX_NUM_SGE;
+	qp_init_attr.cap.max_recv_sge = MCE_DEFAULT_RX_NUM_SGE;
+	qp_init_attr.sq_sig_all = 0;
+	qp_init_attr.qp_type = qp_type;
+
+	//find ib_cxt
+	char base_ifname[IFNAMSIZ];
+	get_base_interface_name((const char*)(ifname), base_ifname, sizeof(base_ifname));
+	struct ibv_context** pp_ibv_context_list = rdma_get_devices(&num_devices);
+	char resource_path[256];
+	sprintf(resource_path, VERBS_DEVICE_RESOURCE_PARAM_FILE, base_ifname);
+	char sys_res[1024] = {0};
+	priv_safe_read_file(resource_path, sys_res, 1024);
+	for (int j=0; j<num_devices; j++) {
+		char ib_res[1024] = {0};
+		char ib_path[256] = {0};
+		sprintf(ib_path, "%s/device/resource", pp_ibv_context_list[j]->device->ibdev_path);
+		priv_safe_read_file(ib_path, ib_res, 1024);
+		if (strcmp(sys_res, ib_res) == 0) {
+			//create qp resources
+			ib_ctx_handler* p_ib_ctx = g_p_ib_ctx_handler_collection->get_ib_ctx(pp_ibv_context_list[j]);
+			struct ibv_comp_channel *channel = ibv_create_comp_channel(p_ib_ctx->get_ibv_context());
+			if (!channel) {
+				ndtm_logdbg("channel creation failed for interface %s (errno=%d %m)", ifname, errno);
+				break;
+			}
+			struct ibv_cq* cq = vma_ibv_create_cq(p_ib_ctx->get_ibv_context(), safe_mce_sys().tx_num_wr, (void*)this, channel, 0, &attr);
+			if (!cq) {
+				ndtm_logdbg("cq creation failed for interface %s (errno=%d %m)", ifname, errno);
+				break;
+			}
+			qp_init_attr.recv_cq = cq;
+			qp_init_attr.send_cq = cq;
+			struct ibv_qp* qp = ibv_create_qp(p_ib_ctx->get_ibv_pd(), &qp_init_attr);
+			if (qp) {
+				success = true;
+				IF_VERBS_FAILURE(ibv_destroy_qp(qp)) {
+					ndtm_logdbg("qp destroy failed on interface %s (errno=%d %m)", ifname, errno);
```

Comment:
not sure about this, why shouldn't we offload this interface only beacuse qp destruction failed? if qp was created successfully in my opinion it's good enough to continue